clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3783,6 +3624,323 @@ static void handleSYCLIntelMaxWorkGroupSize(Sema &S, Decl *D,
+    // If any of the results are known to be different, we can diagnose at this
+    // point and drop the attribute.
+    if (llvm::is_contained(Results, DupArgResult::Different)) {
+      Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+      Diag(Existing->getLoc(), diag::note_previous_attribute);
+      return;
+    }
+    // If all of the results are known to be the same, we can silently drop the
+    // attribute. Otherwise, we have to add the attribute and resolve its
+    // differences later.
+    if (llvm::all_of(Results,
+                     [](DupArgResult V) { return V == DupArgResult::Same; }))
+      return;
+  }
+
+  D->addAttr(::new (Context)
+                 ReqdWorkGroupSizeAttr(Context, CI, XDim, YDim, ZDim));
+}
+
+ReqdWorkGroupSizeAttr *
+Sema::MergeReqdWorkGroupSizeAttr(Decl *D, const ReqdWorkGroupSizeAttr &A) {
+  // If the declaration has a ReqdWorkGroupSizeAttr,
+  // check to see if the attribute holds equal values to
+  // (1, 1, 1) in case the value of SYCLIntelMaxGlobalWorkDimAttr
+  // equals to 0.
+  if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxGlobalWorkDimAttr>()) {
+    if (InvalidWorkGroupSizeAttrs(DeclAttr->getValue(), A.getXDim(),
+                                  A.getYDim(), A.getZDim())) {
+      Diag(A.getLoc(), diag::err_sycl_x_y_z_arguments_must_be_one)
+          << &A << DeclAttr;
+      return nullptr;
+    }
+  }
+
+  // If the 'max_work_group_size' attribute is specified on
+  // a declaration along with 'reqd_work_group_size' attribute, check to
+  // see if values of reqd_work_group_size attribute arguments are
+  // equal or less than values of 'max_work_group_size' attribute arguments.
+  //
+  // The arguments to reqd_work_group_size are ordered based on which index
+  // increments the fastest. In OpenCL, the first argument is the index that
+  // increments the fastest, and in SYCL, the last argument is the index that
+  // increments the fastest.
+  //
+  // __attribute__((reqd_work_group_size)) follows the OpenCL rules in OpenCL
+  // mode. All spellings of reqd_work_group_size attribute
+  // (regardless of syntax used) follow the SYCL rules when in SYCL mode.
+  if (const auto *DeclAttr = D->getAttr<SYCLIntelMaxWorkGroupSizeAttr>()) {
+    if (checkWorkGroupSizeAttrValues<std::greater<int>>(
+            *this, A.getXDim(), A.getYDim(), A.getZDim(), DeclAttr->getXDim(),
+            DeclAttr->getYDim(), DeclAttr->getZDim())) {
+      Diag(DeclAttr->getLoc(), diag::err_conflicting_sycl_function_attributes)
+          << DeclAttr << &A;
+      Diag(A.getLoc(), diag::note_conflicting_attribute);
+      return nullptr;
+    }
+  }
+
+  // If the 'reqd_work_group_size' attribute is specified on
+  // a declaration along with 'reqd_work_group_size' attribute,
+  // check to see if values of 'reqd_work_group_size' attribute arguments are
+  // equal or greater than values of 'reqd_work_group_size' attribute arguments.
```

Comment:
```suggestion
  // If the 'reqd_work_group_size' attribute is specified multiple times on
  // a declaration, check if the values of 'reqd_work_group_size' attribute arguments specified earlier are
  // less than the values of 'reqd_work_group_size' attribute arguments specified after.
```
(format as appropriate)

Reply:
I did not make any changes to the comments. I will wait for response from @AaronBallman or other reviewers before making this change.